src/matchers/archive.rs

```diff
@@ -188,9 +190,33 @@ pub fn is_dcm(buf: &[u8]) -> bool {
     buf.len() > 131 && buf[128] == 0x44 && buf[129] == 0x49 && buf[130] == 0x43 && buf[131] == 0x4D
 }
 
+const ZSTD_SKIP_START: usize = 0x184D2A50;
+const ZSTD_SKIP_MASK: usize = 0xFFFFFFF0;
+
 /// Returns whether a buffer is a Zstd archive.
+// Zstandard compressed data is made of one or more frames.
+// There are two frame formats defined by Zstandard: Zstandard frames and Skippable frames.
+// See more details from https://tools.ietf.org/id/draft-kucherawy-dispatch-zstd-00.html#rfc.section.2
 pub fn is_zst(buf: &[u8]) -> bool {
-    buf.len() > 3 && buf[0] == 0x28 && buf[1] == 0xB5 && buf[2] == 0x2F && buf[3] == 0xFD
+    if buf.len() > 3 && buf[0] == 0x28 && buf[1] == 0xB5 && buf[2] == 0x2F && buf[3] == 0xFD {
+        return true;
+    }
+
+    if buf.len() < 8 {
+        return false;
+    }
+
+    let magic = u32::from_le_bytes(buf[0..4].try_into().unwrap()) as usize;
+    if magic & ZSTD_SKIP_MASK == ZSTD_SKIP_START {
+        let data_len = u32::from_le_bytes(buf[4..8].try_into().unwrap()) as usize;
```

Comment:
Converting it into a u32 and then casting it to usize feels very fragile to me, and would definitely break on 16 bit systems (which I have zero knowledge in, but infer could probably run perfectly on them otherwise). It also adds the ambiguity of what happens on bit endian platforms. I think we should compare slices directly.

Reply:
The solution would be to bound-check the u32->usize conversion.


`impl TryFrom<u32> for usize` does exactly what you want, it never fails on 32 and 64-bit machines, but might fail on 16-bit ones.

I see that you've been busy (btw, ghz is pretty cool!) and this PR was hanging for long, so I re-made the PR with the fixes needed, I hope you don't mind.

#88